### PR TITLE
Issue 202

### PR DESCRIPTION
**Pull request by p-alik**
PR aims to solve #202.
Additionally it contains few commits from #114.

### CODE REVIEW

File: libtest/signal.cc [hunk 1]
```diff
@@ -215,7 +216,11 @@ SignalThread::SignalThread() :
   sigaddset(&set, SIGUSR2);
 
   strcpy(lock_name, "/XXXXXXXX");
-  mktemp(lock_name);
+  char *mkdt = mkdtemp(lock_name);
+  if(mkdt == nullptr)
```
Comment (esabol) [inline libtest/signal.cc]: There should be a space between `if` and `(`.

File: util/signal.cc [hunk 2]
```diff
@@ -206,7 +206,12 @@ SignalThread::SignalThread(bool exit_on_signal_arg) :
   sigaddset(&set, SIGUSR2);
 
   strcpy(lock_name, "/XXXXXXXXX");
-  mktemp(lock_name);
+  char *mkdt = mkdtemp(lock_name);
+  if(mkdt == nullptr)
```
Comment (esabol) [inline util/signal.cc]: There should be a space between `if` and `(`.

**Comment by esabol** [comment]
Clearly, the code works, but I'm trying to figure out how it works when one doesn't have permissions to create a directory in `/`. Wouldn't mktemp/mkdtemp fail in that scenario? Why doesn't the template need to be `"/tmp/XXXXXXXXX"`?

**Comment by p-alik** [comment]
> Wouldn't mktemp/mkdtemp fail in that scenario? 

Indeed it would.
```
 ./t/memcached 
libtest/signal.cc:222: in SignalThread() pid(15590) Permission denied mkdtemp failed.
memcached.gearmand options.--queue-type=libmemcached --libmemcached-servers=				0:055211511 [ ok ]
memcached.memcached queue.gearman_client_echo()					0:000675759 [ ok ]
memcached.memcached queue.gearman_client_echo() fail					0:000154914 [ ok ]
memcached.memcached queue.gearman_worker_echo()					0:000535230 [ ok ]
memcached.memcached queue.clean					3:001090431 [ ok ]
memcached.memcached queue.add					0:002448225 [ ok ]
memcached.memcached queue.worker					0:001146890 [ ok ]
memcached.queue restart.lp:1054377					3:007742225 [ ok ]
```

> Why doesn't the template need to be "/tmp/XXXXXXXXX"?

I'll try to use proper template.

**Comment by esabol** [comment]
Did mktemp fail in the same scenario? I don't recall getting test failures where compiling as non-root....

Probably should obey TMPDIR environment variable instead of hardcoding /tmp.

**Comment by p-alik** [comment]
@SpamapS added `mktemp` in fb05e80af2bd7817925bb532220b71d3b52cfbef for sake of Mac OS.
It seems the only reason for `mktemp` is to create a random `lock_name` for `sem_open`.
Therefore I'm closing the PR because it is completely inappropriate and wrong.

**Comment by esabol** [comment]
I respectfully disagree and feel this should be reopened, but it is your PR.

I believe `mktemp` will also fail if you run it as non-root. Can you test that? We should not expect users to run tests as root.

I've looked for other code that uses `sem_open` and `mktemp`, and one case I found uses `/tmp/XXXXXX` for the template. So I think we were on the right track....

**Comment by p-alik** [comment]
> I believe `mktemp` will also fail if you run it as non-root. Can you test that?

`mktemp` doesn't fail.

```c++
#include <iostream>
#include <stdlib.h>

int main(void) {
  char tmpl[] = "/XXXXXXXX";
  mktemp(tmpl);
  std::cout << "tmpl =" << tmpl << " after mktemp('/XXXXXXXX') " << std::endl;

  int fd = mkstemp(tmpl);
  if (fd == -1) {
    std::cout << "mkstemp('" << tmpl << "') failed " << std::endl;
    return 1;
  }
}
```

```
$ ./a.out 
tmpl =/XXXN51hx after mktemp('/XXXXXXXX') 
mkstemp('/XXXN51hx') failed
```

**Comment by esabol** [comment]
Ah, I see! `mktemp` doesn't actually create the file. It just checks to see if the file exists. Duh. Sorry. Too early in the morning for me.

Btw, your test program isn't quite correct. `tmpl` needs to be reset between calling `mktemp` and `mkstemp`. `tmpl` needs to have trailing `'XXXXXX'` when you call `mkstemp`. But it doesn't matter.

Anyway, glibc's implementation of pthread eliminated the same warning message by changing `mktemp` to `__mktemp`. What do you think of doing the following just to get rid of the warning?

```C
#ifdef __GLIBC__
#define mktemp(X) __mktemp(X)
#endif
```

**Comment by p-alik** [comment]
I couldn't reopen the PR. The only difference in #205:
- `mkstemp` replaces `mktemp`
- the result of `mkstemp` will be ignored

**Pull request closed without merging**